Cabal/tests/PackageTests/PackageTester.hs

```diff
@@ -1,367 +1,540 @@
+    return Result {
+            resultExitCode = exitcode,
+            resultDirectory = mb_cwd,
+            resultCommand = fullCmd,
+            resultOutput = out
+        }
 
 ------------------------------------------------------------------------
--- * Test helpers
-
-assertConfigureSucceeded :: Result -> Assertion
-assertConfigureSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup configure\' should succeed\n" ++
-    "  output: " ++ outputText result
-
-assertBuildSucceeded :: Result -> Assertion
-assertBuildSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup build\' should succeed\n" ++
-    "  output: " ++ outputText result
-
-assertBuildFailed :: Result -> Assertion
-assertBuildFailed result = when (successful result) $
-    assertFailure $
-    "expected: \'setup build\' should fail\n" ++
-    "  output: " ++ outputText result
-
-assertHaddockSucceeded :: Result -> Assertion
-assertHaddockSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup haddock\' should succeed\n" ++
-    "  output: " ++ outputText result
-
-assertTestSucceeded :: Result -> Assertion
-assertTestSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup test\' should succeed\n" ++
-    "  output: " ++ outputText result
-
-assertTestFailed :: Result -> Assertion
-assertTestFailed result = when (successful result) $
-    assertFailure $
-    "expected: \'setup test\' should fail\n" ++
-    "  output: " ++ outputText result
+-- * Subprocess run results
 
-assertInstallSucceeded :: Result -> Assertion
-assertInstallSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup install\' should succeed\n" ++
-    "  output: " ++ outputText result
+data Result = Result
+    { resultExitCode :: ExitCode
+    , resultDirectory :: Maybe FilePath
+    , resultCommand :: String
+    , resultOutput :: String
+    } deriving Show
 
-assertRegisterSucceeded :: Result -> Assertion
-assertRegisterSucceeded result = unless (successful result) $
-    assertFailure $
-    "expected: \'setup register\' should succeed\n" ++
-    "  output: " ++ outputText result
+requireSuccess :: Result -> TestM Result
+requireSuccess r@Result { resultCommand = cmd
+                        , resultExitCode = exitCode
+                        , resultOutput = output } = do
+    (_, test) <- ask
+    let lastN n = unlines . reverse . take n . reverse . lines
+    when (exitCode /= ExitSuccess && not (testShouldFail test)) $
+        assertFailure $ "Command " ++ cmd ++ " failed.\n" ++
+        "last 5 lines of output (see test.log for full):\n" ++ lastN 5 output
+    when (exitCode == ExitSuccess && testShouldFail test) $
+        assertFailure $ "Command " ++ cmd ++ " succeeded.\n" ++
+        "last 5 lines of output (see test.log for full):\n" ++ lastN 5 output
+    -- TODO: breadcrumb for test.log for full output
```

Comment:
Just show all the output. The last N lines are *almost always* the *most* uninteresting. (We've removed the "tail N" effect in regular builds... unless it's been re-enabled since I removed it ~6-12 months or so ago.)